### PR TITLE
deps: chakrashim enable JavaScript debugging

### PR DESCRIPTION
**Pull request by agarwal-sandeep**
##### Checklist

- [x] `make -j4 test` (UNIX), or `vcbuild test nosign` (Windows) passes
- [x] commit message follows commit guidelines
##### Affected core subsystem(s)

chakrashim, node
##### Description of change

This change adds JavaScript debugging support in node-chakracore by using
JsRT debugging apis added in chakracore release 1.3. A shim convert v8
debugging protocol messages to chakracore debugging apis and visa-versa.

node-chakracore only expose the debugging protocol messages and doesn't
shim all the apis which v8 provides, so it is only debuggable using tools
which just relies on debugging protocol json and doesn't use internals
of v8 debugging implementations.

PR-URL: nodejs/node-chakracore#[To be filled]
Reviewed-By: [To be filled]

### CODE REVIEW

File: deps/chakrashim/lib/chakra_debug.js [hunk 1]
```diff
@@ -0,0 +1,1304 @@
...
+  function AddChildrens(obj) {
+    if ('display' in obj) {
+      if (!('value' in obj) || !isFinite(obj['value']) ||
+        obj['value'] == null || obj['value'] == undefined) {
+        obj['value'] = obj['display'];
+      }
+      if (!('text' in obj)) {
+        obj['text'] = obj['display'];
+      }
+      if (!('className' in obj)) {
```
Comment (digitalinfinity) [inline deps/chakrashim/lib/chakra_debug.js:127]: Curious- why is this within the 'display' in obj case? Should this be in 'type' in obj? Also, curious, do you need to use in here or would hasOwnProperty suffice? #Resolved

File: deps/chakrashim/lib/chakra_debug.js [hunk 2]
```diff
@@ -0,0 +1,1304 @@
...
+      obj.protoObject = {
+        'ref': globalExecutionState.GetObjectProtoEvalHandle()
+      };
+      obj.prototypeObject = {
+        'ref': globalExecutionState.GetObjectPrototypeEvalHandle()
+      };
+    }
+
+    var PROPERTY_ATTRIBUTE_HAVE_CHILDRENS = 0x1;
+    //var PROPERTY_ATTRIBUTE_READ_ONLY_VALUE = 0x2;
```
Comment (digitalinfinity) [inline deps/chakrashim/lib/chakra_debug.js:145]: Nit: remove commented code? #WontFix

File: deps/chakrashim/lib/chakra_debug.js [hunk 3]
```diff
@@ -0,0 +1,1304 @@
...
+          Logger.LogError('ProcessPendingMessages exception: ' + ex.message,
+                          ex.stack);
+        }
+      }
+    };
+  }
+
+  ChakraDebugEventProcessor.prototype[ChakraDebugEvent.SourceCompile] =
+    function() {
+    // {"fileName":"d:\\test.js","lineCount":4,"sourceLength":113,"scriptId":2}
```
Comment (digitalinfinity) [inline deps/chakrashim/lib/chakra_debug.js]: remove commented out line? #Resolved

File: deps/chakrashim/src/v8debug.cc [hunk 4]
```diff
@@ -21,71 +21,715 @@
...
-      JsStartDebugging();
+  void CALLBACK JsDiagDebugEventHandler(
+    JsDiagDebugEvent debugEvent,
+    JsValueRef eventData,
+    void* callbackState) {
+    if (Debug::handler == nullptr) {
+      return;
     }
+
+    Debug::messageQueue.debugEventProcessCount++;
```
Comment (digitalinfinity) [inline deps/chakrashim/src/v8debug.cc]: Is this always called on the same thread? If not, should this be interlocked? #ByDesign

File: deps/chakrashim/src/v8debug.cc [hunk 5]
```diff
@@ -21,71 +21,715 @@
...
+
+  MessageQueue::~MessageQueue() {
+    uv_mutex_destroy(&msgMutex);
+    uv_sem_destroy(&newMsgSem);
+    delete this->msgQueue;
+    this->msgQueue = nullptr;
+  }
+
+  void MessageQueue::SaveMessage(const uint16_t* command, int length) {
+    uint16_t* msg = new uint16_t[length + 1];
```
Comment (digitalinfinity) [inline deps/chakrashim/src/v8debug.cc]: Where is this msg deleted? Might be good to add a comment here on the lifetime semantics of this allocation. #Resolved

**Review by digitalinfinity — APPROVED** [review]
Did a cosmetic review since I'm assuming this was reviewed when the changes where made in the debugging branch- overall, LGTM

**Comment by agarwal-sandeep** [comment]
@digitalinfinity This changes were not reviewed previously

File: deps/chakrashim/lib/chakra_debug.js (hunk 2 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/lib/chakra_debug.js:145]: Currently we don't use the read only attribute, but I will keep it here for clarity.

---

In reply to: [81491126](https://github.com/nodejs/node-chakracore/pull/126#discussion_r81491126) [](ancestors = 81491126)

File: deps/chakrashim/src/v8debug.cc (hunk 4 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/src/v8debug.cc]: JsDiagDebugEventHandler is supposed to be per runtime, so it is only called from one runtime at a time. No need for interlock.

---

In reply to: [81492243](https://github.com/nodejs/node-chakracore/pull/126#discussion_r81492243) [](ancestors = 81492243)

File: deps/chakrashim/lib/chakra_debug.js [hunk 6]
```diff
@@ -0,0 +1,1304 @@
...
+    function GetNextInternalHandle() {
+      return --_internalHandle;
+    }
+
+    var _nextResponseSeq = 0;
+
+    function Response(request) {
+      this.seq = _nextResponseSeq++;
+      this.type = 'response';
+      this.request_seq = request.seq;
```
Comment (obastemur) [inline deps/chakrashim/lib/chakra_debug.js:182]: `this.request_seq` unused property ? #ByDesign

File: deps/chakrashim/lib/chakra_debug.js [hunk 7]
```diff
@@ -0,0 +1,1304 @@
...
+            }
+          }
+        }
+
+        if (filter != null) {
+          found = false;
+          if (filter == element.GetFileName()) {
+            found = true;
+          }
+        }
```
Comment (obastemur) [inline deps/chakrashim/lib/chakra_debug.js:655]: from `652` to `655` can be simplified into one line

found = filter == element.GetFileName(); #Resolved

File: deps/chakrashim/src/jsrtcontextshim.cc [hunk 8]
```diff
@@ -447,6 +451,46 @@ bool ContextShim::ExecuteChakraShimJS() {
...
+    buffer,
+    _countof(chakra_debug_native)) != JsNoError) {
+    return false;
+  }
+
+  // Ensure the buffer is null terminated
+  buffer[_countof(chakra_debug_native)] = L'\0';
+
+  JsValueRef getInitFunction;
+  if (JsParseScriptWithAttributes(buffer,
```
Comment (obastemur) [inline deps/chakrashim/src/jsrtcontextshim.cc:469]: I'm not sure if this works on xplat. Reminder: [unix wchar_t 4 bytes] #Resolved

File: deps/chakrashim/src/v8debug.cc [hunk 9]
```diff
@@ -21,71 +21,715 @@
...
+  }
+
+  MessageQueue::~MessageQueue() {
+    uv_mutex_destroy(&msgMutex);
+    uv_sem_destroy(&newMsgSem);
+    delete this->msgQueue;
+    this->msgQueue = nullptr;
+  }
+
+  void MessageQueue::SaveMessage(const uint16_t* command, int length) {
```
Comment (obastemur) [inline deps/chakrashim/src/v8debug.cc]: `uint16_t` and `wchar_t` ? (2 bytes / 4 bytes) #Resolved

File: deps/chakrashim/src/v8debug.cc [hunk 10]
```diff
@@ -21,71 +21,715 @@
...
+
+  bool MessageQueue::ShouldContinue(JsDiagDebugEvent debugEvent) {
+    JsValueRef debugEventRef;
+    JsErrorCode errorCode = JsIntToNumber(debugEvent, &debugEventRef);
+    CHAKRA_VERIFY_NOERROR(errorCode);
+
+    JsValueRef args[] = { jsrt::GetUndefined(), debugEventRef };
+    JsValueRef result;
+    errorCode = JsCallFunction(
+      this->processShouldContinueFn, args, _countof(args), &result);
```
Comment (obastemur) [inline deps/chakrashim/src/v8debug.cc]: Do we have a definition for `_countof` non-pal ? #Resolved

File: deps/chakrashim/lib/chakra_debug.js (hunk 6 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/lib/chakra_debug.js:182]: This is part of debugging protocol [https://github.com/v8/v8/wiki/Debugging-Protocol](https://github.com/v8/v8/wiki/Debugging-Protocol)

---

In reply to: [81944574](https://github.com/nodejs/node-chakracore/pull/126#discussion_r81944574) [](ancestors = 81944574)

File: src/node.cc [hunk 11]
```diff
@@ -4250,6 +4245,13 @@ void Init(int* argc,
     exit(9);
   }
 
+  // CHAKRA-TODO : fix this to not do it here
```
Comment (kunalspathak) [inline src/node.cc:4248]: > HAKRA-TODO : fix this to not do it here [](start = 6, length = 39)

What is the problem here? #ByDesign

File: deps/chakrashim/src/jsrtisolateshim.cc [hunk 12]
```diff
@@ -59,12 +63,6 @@ IsolateShim::~IsolateShim() {
 }
 
 /* static */ v8::Isolate * IsolateShim::New() {
-  // CHAKRA-TODO: Disable multiple isolate for now until it is fully implemented
```
Comment (kunalspathak) [inline deps/chakrashim/src/jsrtisolateshim.cc:62]: > CHAKRA-TODO: Disable multiple isolate for now until it is fully implemented [](start = 5, length = 75)

Has this changed now? Why is this removed? #ByDesign

File: deps/chakrashim/src/jsrtcontextshim.cc [hunk 13]
```diff
@@ -447,6 +451,46 @@ bool ContextShim::ExecuteChakraShimJS() {
...
+  JsBoolToBoolean(v8::g_trace_debug_json, &traceDebugJsonRef);
+
+  JsValueRef arguments[] = { this->globalObject, this->globalObject,
+      this->keepAliveObject, traceDebugJsonRef };
+  JsErrorCode errorCode = JsCallFunction(initFunction, arguments,
+      _countof(arguments), chakraDebugObject);
+
+  CHAKRA_VERIFY_NOERROR(errorCode);
+
+  return errorCode == JsNoError;
```
Comment (kunalspathak) [inline deps/chakrashim/src/jsrtcontextshim.cc]: > rrorCode == JsNoError; [](start = 10, length = 22)

nit: won't this always be true given that you have CHAKRA_VERIFY_NOERROR(errorCode); above? #Resolved

File: deps/chakrashim/src/jsrtcontextshim.cc [hunk 14]
```diff
@@ -447,6 +451,46 @@ bool ContextShim::ExecuteChakraShimJS() {
                         &result) == JsNoError;
 }
 
+bool ContextShim::ExecuteChakraDebugShimJS(JsValueRef * chakraDebugObject) {
```
Comment (kunalspathak) [inline deps/chakrashim/src/jsrtcontextshim.cc:455]: > ExecuteChakraDebugShimJS [](start = 18, length = 24)

suggestion : can we extract out common code of ExecuteChakraDebugShimJS and ExecuteChakraShimJS because in future we might have additional shim.js files that we will execute? #WontFix

File: deps/chakrashim/src/v8context.cc [hunk 15]
```diff
@@ -71,7 +72,34 @@ Local<Context> Context::New(Isolate* external_isolate,
     return Local<Context>();
   }
 
-  return Local<Context>::New(external_isolate, static_cast<Context *>(context));
+  Local<Context> thisContext = Local<Context>::New(
+      external_isolate, static_cast<Context *>(context));
+
+  if (v8::Debug::IsDebugEnabled() && isoShim->debugContext == nullptr) {
+    // If JavaScript debugging APIs need to be exposed then make sure
+    // debugContext is avaliable and chakra_debug.js is compiled. Inject
```
Comment (kunalspathak) [inline deps/chakrashim/src/v8context.cc]: > avaliable [](start = 23, length = 9)

nit: spelling #Resolved

File: deps/chakrashim/src/v8debug.cc [hunk 16]
```diff
@@ -21,71 +21,715 @@
...
+    // commands on JsDiagDebugEventAsyncBreak
+    JsErrorCode errorCode = JsDiagRequestAsyncBreak(
+      jsrt::IsolateShim::FromIsolate(isolate)->GetRuntimeHandle());
+    CHAKRA_VERIFY_NOERROR(errorCode);
   }
 
-error:
-#else
-  hr = E_FAIL;  // ChakraCore does not support JsStartDebugging
+  static JsValueRef CALLBACK Log(
```
Comment (kunalspathak) [inline deps/chakrashim/src/v8debug.cc]: > Log [](start = 29, length = 3)

If i understand correctly is this for debugging the debugger code and will never called from user scenario?  #ByDesign

File: deps/chakrashim/src/v8debug.cc [hunk 17]
```diff
@@ -21,71 +21,715 @@
...
+      success = true;
+    }
+
+    JsValueRef returnRef = JS_INVALID_REFERENCE;
+    JsBoolToBoolean(success, &returnRef);
+
+    return returnRef;
+  }
+
+  static JsValueRef CALLBACK JsSetBreakpoint(
```
Comment (kunalspathak) [inline deps/chakrashim/src/v8debug.cc]: > JsSetBreakpoint [](start = 29, length = 15)

Can you different nomenclature for these methods? Usually Js*() methods are Jsrt APIs. Would be good to reflect that these are shim methods. #WontFix

File: deps/chakrashim/src/v8debug.cc (hunk 16 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/src/v8debug.cc]: Yes #Resolved

File: deps/chakrashim/src/v8debug.cc (hunk 17 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/src/v8debug.cc]: all jsrt debugging APIs are named JsDiag\* so I choose this name, would keep it this way. #Closed

File: deps/chakrashim/src/jsrtisolateshim.cc (hunk 12 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/src/jsrtisolateshim.cc:62]: Because now we need atleast 2 isolate (one main isolate and another for debugger agent). #Resolved

File: src/node.cc (hunk 11 above)
Comment (agarwal-sandeep) [reply src/node.cc:4248]: No problem, but I would like to find a way to not make any NODE_ENGINE_CHAKRACORE specific change to node.cc and somehow determine this in shim itself #ByDesign

File: deps/chakrashim/lib/chakra_debug.js [hunk 18]
```diff
@@ -0,0 +1,1304 @@
...
+        scopeAndRef.scope = {
+          'type': scopeType,
+          'index': index,
+          'frameIndex': frameIndex,
+          'object': {
+            'ref': scopeRef
+          }
+        };
+        var properties = [];
+        for (var propsLen = 0; propsLen < propertiesArray.length; ++propsLen) {
```
Comment (kunalspathak) [inline deps/chakrashim/lib/chakra_debug.js]: > for (va [](start = 8, length = 7)

suggestion : some of the for-loops and foreach can be converted to .map() #Resolved

File: deps/chakrashim/src/jsrtcontextshim.cc (hunk 8 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/src/jsrtcontextshim.cc:469]: Debugging support is not xplat ready yet, once we have xplat string APIs we need to use that and make some changes. Take a look at https://github.com/agarwal-sandeep/node-chakracore/blob/debugging-vscode-xplat-jc/deps/chakrashim/src/v8debug.cc which uses the prototype xplat string APIs. #Resolved

File: deps/chakrashim/lib/chakra_debug.js [hunk 19]
```diff
@@ -0,0 +1,1304 @@
...
+
+      this.frames = _frames;
+    }
+    return this.frames;
+  };
+
+  ExecutionState.prototype.GetObjectEvalHandle = function() {
+    if (!this.objectEval) {
+      var objEval = CallHostFunction(chakraDebug.JsDiagEvaluate, 'Object', 0);
+      this.objectEval = objEval[1];
```
Comment (kunalspathak) [inline deps/chakrashim/lib/chakra_debug.js:408]: > bjEval[1] [](start = 25, length = 9)

If i understand correctly, objEval[0] holds if call succeed or not. You never check that? #Resolved

File: deps/chakrashim/src/v8debug.cc (hunk 9 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/src/v8debug.cc]: On windows uint16_t and wchar_t are both 2 bytes and debugging support is currently only for windows. I have a static_assert to make sure we fix this for xplat #Resolved

File: deps/chakrashim/lib/chakra_debug.js (hunk 19 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/lib/chakra_debug.js:408]: Should never fail as we are trying to get the global 'Object' #Resolved

File: deps/chakrashim/lib/chakra_debug.js [hunk 20]
```diff
@@ -0,0 +1,1304 @@
...
+
+    if (breakpointScript) {
+      var v8Breakpoint = new V8Breakpoint(breakpointType, breakpointTarget,
+        breakpointScript, line, column);
+      var success = false;
+      if (v8Breakpoint.Set()) {
+        success = true;
+        DebugManager.BreakpointManager.AddBreakpoint(v8Breakpoint);
+      }
+      // ToDo: Add to list
```
Comment (kunalspathak) [inline deps/chakrashim/lib/chakra_debug.js]: > / ToDo: Add to list [](start = 7, length = 19)

Is this still a TODO? #Resolved

File: deps/chakrashim/src/v8v8.cc [hunk 21]
```diff
@@ -106,11 +106,18 @@ void V8::SetFlagsFromCommandLine(int *argc, char **argv, bool remove_flags) {
...
+      if (remove_flags) {
+        argv[i] = nullptr;
+      }
     } else if (remove_flags &&
                (startsWith(
                  arg, "--debug")  // Ignore some flags to reduce unit test noise
                 || startsWith(arg, "--harmony")
-                || startsWith(arg, "--stack-size="))) {
+                || startsWith(arg, "--stack-size=")
+                || startsWith(arg, "--nolazy"))) {
```
Comment (kunalspathak) [inline deps/chakrashim/src/v8v8.cc:120]: > "--nolazy") [](start = 35, length = 11)

why this change? #ByDesign

Comment (agarwal-sandeep) [reply deps/chakrashim/src/v8v8.cc:120]: Its like disabling deferred parsing in V8, VSCode by defaults add this flag so we need to ignore it. #ByDesign

File: deps/chakrashim/lib/chakra_debug.js (hunk 19 above)
Comment (kunalspathak) [reply deps/chakrashim/lib/chakra_debug.js:408]: Ok. Can we have some an if-check saying if it fails then something bad is going on? #Resolved

File: deps/chakrashim/include/v8-debug.h [hunk 22]
```diff
@@ -26,11 +26,46 @@
 // OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 #pragma once
+
 #include <v8.h>
+#include "uv.h"
+#include <queue>
```
Comment (jianchun) [inline deps/chakrashim/include/v8-debug.h]: nit: cpplint demands <...> before "..." includes. #Resolved

File: deps/chakrashim/include/v8-debug.h [hunk 23]
```diff
@@ -26,11 +26,46 @@
...
 
 namespace v8 {
+class V8_EXPORT MessageQueue {
+ public:
+    MessageQueue();
```
Comment (jianchun) [inline deps/chakrashim/include/v8-debug.h]: nit: cpplint / consistency prefers these to be 2-space indented. #Pending

File: deps/chakrashim/include/v8-debug.h (hunk 22 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/include/v8-debug.h]: strange I ran cpplint and didn't see any warning/error. I will run again. #Resolved

File: deps/chakrashim/include/v8-debug.h [hunk 24]
```diff
@@ -46,25 +81,40 @@ class V8_EXPORT Debug {
...
   static void SendCommand(Isolate* isolate,
                           const uint16_t* command, int length,
-                          ClientData* client_data = NULL) {
-  }
+                          ClientData* client_data = NULL);
   static MaybeLocal<Value> GetMirror(Local<Context> context,
                                      Handle<Value> obj) {
     return MaybeLocal<Value>();
   }
+  static void StartDebugging(JsRuntimeHandle runtime);
```
Comment (jianchun) [inline deps/chakrashim/include/v8-debug.h]: I recall v8::Debug is a public class in v8 headers. These look like deviate from v8 signature. Do we need to put them here, or can we move them to private headers?

Same question to other new classes in this file. #Resolved

File: deps/chakrashim/include/v8-debug.h (hunk 22 above)
Comment (jianchun) [reply deps/chakrashim/include/v8-debug.h]: That's from my memory...

---

In reply to: [82239941](https://github.com/nodejs/node-chakracore/pull/126#discussion_r82239941) [](ancestors = 82239941)

File: deps/chakrashim/lib/chakra_debug.js [hunk 25]
```diff
@@ -0,0 +1,1304 @@
...
+      Output: 2,
+      Input: 3,
+      Error: 4
+    };
+
+    var enabledLogType = [];
+    enabledLogType[logTypes.API] = false; // For development purpose
+    enabledLogType[logTypes.Output] = traceDebugJson === true;
+    enabledLogType[logTypes.Input] = traceDebugJson === true;
+    enabledLogType[logTypes.Error] = false; // For development purpose
```
Comment (jianchun) [inline deps/chakrashim/lib/chakra_debug.js]: nit nit pick: overkill, 4 flag variables suffice #Resolved

Comment (jianchun) [reply deps/chakrashim/lib/chakra_debug.js]: Maybe you can also support turning these on through environment variables, so we don't need to recompile.

---

In reply to: [82242030](https://github.com/nodejs/node-chakracore/pull/126#discussion_r82242030) [](ancestors = 82242030)

File: deps/chakrashim/lib/chakra_debug.js [hunk 26]
```diff
@@ -0,0 +1,1304 @@
...
+    });
+  }
+
+  var Debug = globalObject.Debug;
+
+  Debug.ScriptBreakPointType = {
+    ScriptId: 0,
+    ScriptName: 1,
+    ScriptRegExp: 2
+  };
```
Comment (jianchun) [inline deps/chakrashim/lib/chakra_debug.js:93]: Can these be messed up by user script? Should they be hidden/readonly somehow? #ByDesign

File: deps/chakrashim/lib/chakra_debug.js [hunk 27]
```diff
@@ -0,0 +1,1304 @@
...
+  function ExecutionState(stateType, scriptId) {
+    this.type = stateType;
+    this.breakScriptId = scriptId;
+    this.frames = undefined;
+    this.objectEval = undefined;
+    this.objectProtoEval = undefined;
+    this.objectPrototypeEval = undefined;
+  }
+
+  ExecutionState.prototype.ChangeToBreak = function() {
```
Comment (jianchun) [inline deps/chakrashim/lib/chakra_debug.js:377]: Why not use cool ES6 class for these #ByDesign

File: deps/chakrashim/lib/chakra_debug.js [hunk 28]
```diff
@@ -0,0 +1,1304 @@
...
+        if (filter != null) {
+          found = false;
+          if (filter == element.GetFileName()) {
+            found = true;
+          }
+        }
+
+        if (found == true) {
+          var scriptObj = element.GetProtocolObject();
+          if (includeSource == true) {
```
Comment (jianchun) [inline deps/chakrashim/lib/chakra_debug.js]: nit: if (found) ... if (includeSource) ... #Resolved

File: deps/chakrashim/src/v8v8.cc [hunk 29]
```diff
@@ -140,23 +147,14 @@ bool V8::Initialize() {
   if (g_disposed) {
     return false;  // Can no longer Initialize if Disposed
   }
-#ifndef NODE_ENGINE_CHAKRACORE
```
Comment (jianchun) [inline deps/chakrashim/src/v8v8.cc:143]: What's "node-chakra" / "node-uwp" debugging story after this change? Can this be merged to those branches? (Would this break VS debugging?) #Pending

File: deps/chakrashim/src/v8debug.cc (hunk 10 above)
Comment (jianchun) [reply deps/chakrashim/src/v8debug.cc]: We probably have one defined in xplat branch.

---

In reply to: [81947919](https://github.com/nodejs/node-chakracore/pull/126#discussion_r81947919) [](ancestors = 81947919)

File: deps/chakrashim/src/v8debug.cc (hunk 9 above)
Comment (jianchun) [reply deps/chakrashim/src/v8debug.cc]: Is there reason to use 2 different types? (Why not avoid wchar_t and only use uint16_t and use v8::String where string is needed.)

---

In reply to: [82074191](https://github.com/nodejs/node-chakracore/pull/126#discussion_r82074191) [](ancestors = 82074191)

File: deps/chakrashim/src/v8v8.cc (hunk 29 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/src/v8v8.cc:143]: Disucssed with @digitalinfinity @aruneshchandra and for now we will go ahead with this break and understand why node-chakra/node-uwp needs to take changes from node-chakracore. If we need to support PDM based debugging will bring back in a separate change 

---

In reply to: [82250358](https://github.com/nodejs/node-chakracore/pull/126#discussion_r82250358) [](ancestors = 82250358)

File: deps/chakrashim/lib/chakra_debug.js (hunk 28 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/lib/chakra_debug.js]: Sorry I didn't understood what you mean here? found is used to track if a particular scriptid is requested and includeSource is tracking if additionally we need to include completed source of the script or not.

---

In reply to: [82246914](https://github.com/nodejs/node-chakracore/pull/126#discussion_r82246914) [](ancestors = 82246914)

File: deps/chakrashim/lib/chakra_debug.js (hunk 27 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/lib/chakra_debug.js:377]: Most of these is trying to mimic how its done in V8 (debug.js) and I would like to keep it as close as possible in terms of structure.

---

In reply to: [82246479](https://github.com/nodejs/node-chakracore/pull/126#discussion_r82246479) [](ancestors = 82246479)

File: deps/chakrashim/lib/chakra_debug.js (hunk 28 above)
Comment (kunalspathak) [reply deps/chakrashim/lib/chakra_debug.js]: I think what @jianchun meant is instead of having `if(found == true)` you could simply do `if(found)` and vice versa for `includeSource`. #Resolved

File: deps/chakrashim/lib/chakra_debug.js (hunk 26 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/lib/chakra_debug.js:93]: Yes if you use vm.runInDebugContext and that's by design and same as v8

---

In reply to: [82244119](https://github.com/nodejs/node-chakracore/pull/126#discussion_r82244119) [](ancestors = 82244119)

File: deps/chakrashim/lib/chakra_debug.js (hunk 28 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/lib/chakra_debug.js]: That would make sense, thanks

---

In reply to: [82484091](https://github.com/nodejs/node-chakracore/pull/126#discussion_r82484091) [](ancestors = 82484091)

File: deps/chakrashim/include/v8-debug.h (hunk 22 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/include/v8-debug.h]: Made the change but cpplint.py is not showing at as a error/warning

---

In reply to: [82240850](https://github.com/nodejs/node-chakracore/pull/126#discussion_r82240850) [](ancestors = 82240850,82239941)

File: deps/chakrashim/include/v8-debug.h (hunk 23 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/include/v8-debug.h]: Strange, cpplint doesn't show any error/warning with current spacing but if I make it 2 space cpplint is giving error.

---

In reply to: [82239807](https://github.com/nodejs/node-chakracore/pull/126#discussion_r82239807) [](ancestors = 82239807)

File: deps/chakrashim/src/v8debug.cc (hunk 9 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/src/v8debug.cc]: Thanks! removed wchar_t usage and instead used v8::String::NewFromTwoByte

---

In reply to: [82251507](https://github.com/nodejs/node-chakracore/pull/126#discussion_r82251507) [](ancestors = 82251507,82074191)

File: deps/chakrashim/src/jsrtcontextshim.cc (hunk 14 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/src/jsrtcontextshim.cc:455]: I though of that but that will make things a bit complicated as we need to allocate buffer using new and then in all failure cases make sure it is deleted, Also there not not much command code. We call different APIs in both functions. For now I think we used keep it as it is and if we add more shim then we should definitely refactor this.

---

In reply to: [82066039](https://github.com/nodejs/node-chakracore/pull/126#discussion_r82066039) [](ancestors = 82066039)

File: deps/chakrashim/lib/chakra_debug.js (hunk 20 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/lib/chakra_debug.js]: No, seems a left over. Removed

---

In reply to: [82074541](https://github.com/nodejs/node-chakracore/pull/126#discussion_r82074541) [](ancestors = 82074541)

File: deps/chakrashim/lib/chakra_debug.js (hunk 1 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/lib/chakra_debug.js:127]: If you look at JsrtDebugUtils::AddPropertyType, we always add "type" but className in  only for few cases. This adding of className as type is to satisfy VSCode.
hasOwnProperty will do but it seems most of the place we use in to check for property. 

---

In reply to: [81491092](https://github.com/nodejs/node-chakracore/pull/126#discussion_r81491092) [](ancestors = 81491092)

File: deps/chakrashim/include/v8-debug.h (hunk 24 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/include/v8-debug.h]: moved most of new code to jsrtdebug.cc/jsrtdebug.h

---

In reply to: [82240689](https://github.com/nodejs/node-chakracore/pull/126#discussion_r82240689) [](ancestors = 82240689)

File: deps/chakrashim/lib/chakra_debug.js (hunk 25 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/lib/chakra_debug.js]: Now we only have LogAPI, DebugJson and Error. DebugJson is enabled using --trace_debug_json and Error is enabled by default.

Seems like when we execute chakra_debug process object is not yet set so can't use that to get the environment variable. Will see if I can do something for dynamically enable/disable after the PR as a separate change.

---

In reply to: [82242030](https://github.com/nodejs/node-chakracore/pull/126#discussion_r82242030) [](ancestors = 82242030)

File: deps/chakrashim/lib/chakra_debug.js [hunk 30]
```diff
@@ -0,0 +1,1289 @@
...
+  function CallHostFunctionNoLog(fn) {
+    var args = [].slice.call(arguments, 1);
+    var start = new Date();
+    var result = fn.apply(undefined, args);
+    var end = new Date();
+    Logger.LogAPI(`${fn.name}(${args}):${end - start}`);
+    return result;
+  }
+
+  function CallHostFunction(fn) {
```
Comment (digitalinfinity) [inline deps/chakrashim/lib/chakra_debug.js:74]: Nit: This looks very very similar to the previous function- can you combine?  #WontFix

File: deps/chakrashim/lib/chakra_debug.js [hunk 31]
```diff
@@ -0,0 +1,1289 @@
...
+    var start = new Date();
+    var result = fn.apply(undefined, args);
+    var end = new Date();
+    Logger.LogAPI(`${fn.name}(${args}):${end - start}`, result);
+    return result;
+  }
+
+  if (!globalObject.Debug) {
+    Object.defineProperty(globalObject, 'Debug', {
+      value: {}, enumerable: false, configurable: false, writable: false
```
Comment (digitalinfinity) [inline deps/chakrashim/lib/chakra_debug.js]: The property descriptor keys are set to their default values- why not just do globalObject.Debug = {} #Resolved

File: deps/chakrashim/lib/chakra_debug.js [hunk 32]
```diff
@@ -0,0 +1,1289 @@
...
+
+    var BreakpointManager = {
+      GetBreakpoints: function() {
+        var breakpoints = CallHostFunction(chakraDebug.JsDiagGetBreakpoints);
+        var activeBreakpointList = [];
+        breakpoints.forEach(function(element, index, array) {
+          var bpId = element.breakpointId;
+          activeBreakpointList[bpId] = _breakpoints[bpId];
+        });
+        _breakpoints = activeBreakpointList;
```
Comment (digitalinfinity) [inline deps/chakrashim/lib/chakra_debug.js:256]: Nit: can be a little more succinct doing _breakpoints = breakpoints.map(function(element, index, array) { return element.breakpointId; }) #WontFix

File: deps/chakrashim/src/jsrtdebug.h [hunk 33]
```diff
@@ -0,0 +1,85 @@
...
+class Debugger {
+ public:
+  static MessageQueue messageQueue;
+  static v8::Debug::MessageHandler handler;
+  static bool IsDebugEnabled();
+  static void StartDebugging(JsRuntimeHandle runtime);
+  static void SetChakraDebugObject(JsValueRef chakraDebugObject);
+ private:
+  static void InstallHostCallback(JsValueRef chakraDebugObject,
+  const wchar_t *name,
```
Comment (digitalinfinity) [inline deps/chakrashim/src/jsrtdebug.h]: Nit: indent function parameters #Resolved

**Comment by digitalinfinity** [comment]
:shipit:

Left a bunch of nitpicky comments but you can fix after merging

File: deps/chakrashim/src/jsrtdebug.h (hunk 33 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/src/jsrtdebug.h]: good catch, cpplint didn't catch this.

---

In reply to: [83558816](https://github.com/nodejs/node-chakracore/pull/126#discussion_r83558816) [](ancestors = 83558816)

File: deps/chakrashim/lib/chakra_debug.js (hunk 30 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/lib/chakra_debug.js:74]: CallHostFunction dumps the API output/result whereas CallHostFunctionNoLog doesn't, will keep it separate 

---

In reply to: [83558470](https://github.com/nodejs/node-chakracore/pull/126#discussion_r83558470) [](ancestors = 83558470)

File: deps/chakrashim/lib/chakra_debug.js (hunk 32 above)
Comment (agarwal-sandeep) [reply deps/chakrashim/lib/chakra_debug.js:256]: Changing this is breaking from bp removal logic, will keep it this way to not break anything.

---

In reply to: [83558695](https://github.com/nodejs/node-chakracore/pull/126#discussion_r83558695) [](ancestors = 83558695)

File: deps/chakrashim/lib/chakra_debug.js (hunk 30 above)
Comment (kunalspathak) [reply deps/chakrashim/lib/chakra_debug.js:74]: nit: you could still extract the common code while keeping the functionality separate.

**Comment by digitalinfinity** [comment]
Yay! 🎉 🍾 🍰
